### PR TITLE
Initial resource-name code

### PR DESCRIPTION
**Pull request by chrisdunelm**
There are many different scenarios in which resource-names may be used. Only the basics are tested in this PR. More tests will be added in future PRs.

### CODE REVIEW

File: Google.Api.Generator.Tests/ProtoTests/ResourceNames/ResourceNamesResourceNames.cs [hunk 1]
```diff
@@ -0,0 +1,99 @@
+﻿using gax = Google.Api.Gax;
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTests/ResourceNames/ResourceNamesResourceNames.cs]: No copyright - I may just not have noticed this from previous files. Is that still to-do in general?

File: Google.Api.Generator.Tests/ProtoTests/ResourceNames/ResourceNamesResourceNames.cs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        /// <inheritdoc/>
+        public static bool operator ==(SimpleResourceName a, SimpleResourceName b) => ReferenceEquals(a, b) || (a?.Equals(b) ?? false);
+
+        /// <inheritdoc/>
+        public static bool operator !=(SimpleResourceName a, SimpleResourceName b) => !(a == b);
+    }
+
+    public partial class SimpleResource
+    {
+        /// <summary>
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTests/ResourceNames/ResourceNamesResourceNames.cs]: I can't remember what determines whether we do this or just a SimpleResourceName property. Possibly worth just checking, but I suspect you're right :)

File: Google.Api.Generator/Generation/ResourceNamesGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+                OperatorDeclarationSyntax InequalityOperator()
+                {
+                    var a = Parameter(_ctx.Type(def.One.ResourceNameTyp), "a");
+                    var b = Parameter(_ctx.Type(def.One.ResourceNameTyp), "b");
+                    return OperatorMethod(_ctx.Type<bool>(), "!=")(a, b)
+                        .WithBody(Not(Parens(a.Equality(b))))
+                        .WithXmlDoc(XmlDoc.InheritDoc);
+                }
+            }
+
```
Comment (jskeet) [inline Google.Api.Generator/Generation/ResourceNamesGenerator.cs]: Random blank line?

Comment (chrisdunelm) [reply Google.Api.Generator/Generation/ResourceNamesGenerator.cs]: Oops, yes, will fix.

**Pull request merged**